portfolio/src/test/java/com/google/sps/DataServletTest.java

```diff
@@ -0,0 +1,273 @@
+    when(getResponse.getWriter()).thenReturn(printWriter);
+
+    dataServlet.doPost(postRequest, postResponse);
+    dataServlet.doGet(getRequest, getResponse);
+
+    printWriter.flush(); //may not have flushed yet
+    String[] comments = convertStringToArray(stringWriter.toString());
+    assertThat(comments[0]).isEqualTo(COMMENT_ONE);
+    assertThat(comments.length).isEqualTo(1);
+  }
+  
+  @Test
+  public void testPostAndGetSomeComments() throws Exception {
+    int NUM_COMMENTS = 2;
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(getResponse.getWriter()).thenReturn(printWriter);
+
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_ONE);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_TWO);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_THREE);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_FOUR);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_FIVE);
+    dataServlet.doPost(postRequest, postResponse);
+    
+    when(getRequest.getParameter(NUMBER_COMMENTS_DISPLAYED_PARAMETER)).thenReturn(Integer.toString(NUM_COMMENTS));
+    dataServlet.doGet(getRequest, getResponse);
+
+    printWriter.flush(); //may not have flushed yet
+    String[] comments = convertStringToArray(stringWriter.toString());
+    assertThat(comments[0]).isEqualTo(COMMENT_ONE);
+    assertThat(comments[1]).isEqualTo(COMMENT_TWO);
+    assertThat(comments.length).isEqualTo(2);
+  }
+
+  @Test
+  public void testPostAndGetAllComments() throws Exception {
+    int NUM_COMMENTS = 3;
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(getResponse.getWriter()).thenReturn(printWriter);
+
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_ONE);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_TWO);
+    dataServlet.doPost(postRequest, postResponse);
+    when(postRequest.getParameter(COMMENT_PARAMETER)).thenReturn(COMMENT_THREE);
+    dataServlet.doPost(postRequest, postResponse);
+    when(getRequest.getParameter(NUMBER_COMMENTS_DISPLAYED_PARAMETER)).thenReturn(Integer.toString(NUM_COMMENTS));
+    dataServlet.doGet(getRequest, getResponse);
+
+    printWriter.flush(); //may not have flushed yet
+    String[] comments = convertStringToArray(stringWriter.toString());
+    assertThat(comments[0]).isEqualTo(COMMENT_ONE);
+    assertThat(comments[1]).isEqualTo(COMMENT_TWO);
+    assertThat(comments[2]).isEqualTo(COMMENT_THREE);
+    assertThat(comments.length).isEqualTo(3);
```

Comment:
This pattern keeps repeating for most of your tests so you can refactor it into a helper method:

```java
private void verifyReturnedComments(StringWriter writer, String... expectedComments) {
  String[] actualComments = convertStringToArray(writer.toString());
  assertThat(actualComments).asList().containsExactlyElementsIn(expectedComments);
}
```

Reply:
done!